core/src/main/java/org/apache/calcite/rel/rules/DateRangeRules.java

```diff
@@ -401,21 +402,33 @@ private RexNode toRex(RexNode operand, Range<Calendar> r) {
             : SqlStdOperatorTable.GREATER_THAN;
         nodes.add(
             rexBuilder.makeCall(op, operand,
-                rexBuilder.makeDateLiteral(
-                    DateString.fromCalendarFields(r.lowerEndpoint()))));
+                makeLiteralForComparison(rexBuilder, r.lowerEndpoint(), operand)));
       }
       if (r.hasUpperBound()) {
         final SqlBinaryOperator op = r.upperBoundType() == BoundType.CLOSED
             ? SqlStdOperatorTable.LESS_THAN_OR_EQUAL
             : SqlStdOperatorTable.LESS_THAN;
         nodes.add(
             rexBuilder.makeCall(op, operand,
-                rexBuilder.makeDateLiteral(
-                    DateString.fromCalendarFields(r.upperEndpoint()))));
+                makeLiteralForComparison(rexBuilder, r.upperEndpoint(), operand)));
       }
       return RexUtil.composeConjunction(rexBuilder, nodes, false);
     }
 
+    private RexLiteral makeLiteralForComparison(RexBuilder rexBuilder, Calendar calendar,
+        RexNode operand) {
+      switch (operand.getType().getSqlTypeName()) {
+      case TIMESTAMP:
+      case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+        return rexBuilder.makeTimestampLiteral(TimestampString.fromCalendarFields(calendar),
+            operand.getType().getPrecision());
+      case DATE:
+        return rexBuilder.makeDateLiteral(DateString.fromCalendarFields(calendar));
+      default:
+        throw Util.unexpected(operand.getType().getSqlTypeName());
```

Comment:
not sure if raising an exception is the way to go but am not sure about this.

Reply:
I think throwing is right. Fail fast.